### PR TITLE
Add Dockerfile to support running in a container

### PR DESCRIPTION
**Pull request by jbreiding**
Adding Dockerfile for demo purposes with temporal ui

**What changed?**
Adding a Dockerfile for local development

**Why?**
Setting up a local development environment using
https://github.com/jbreiding/potential-goggles

To work on and demo both code bases.

**How did you test it?**
docker and docker-compose
vscode devcontainers
GH codespaces

``` sh
$ docker build -t temporalite:testv2 . --no-cache
$ docker run temporalite:testv2
```

**Potential risks**
None

**Is hotfix candidate?**
No.

### CODE REVIEW

File: internal/liteconfig/config.go [hunk 1]
```diff
@@ -175,12 +177,24 @@ func (o *Config) mustGetService(frontendPortOffset int) config.Service {
...
+		}
+	} else {
+		return config.Service{
+			RPC: config.RPC{
+				GRPCPort:        grpcPort,
+				MembershipPort:  membershipPort,
+				BindOnLocalHost: true,
+				BindOnIP:        "",
+			},
+		}
```
Comment (jlegrone) [inline internal/liteconfig/config.go:198]: Nit: I think this whole function would be a little more clear if we refactored to initialize a service config and then conditionally modify its fields before returning. We could also add some code comments. Eg.

```go
func (o *Config) mustGetService(frontendPortOffset int) config.Service {
	svc := config.Service{
		RPC: config.RPC{
			GRPCPort:        o.FrontendPort + frontendPortOffset
			MembershipPort:  o.FrontendPort + 100 + frontendPortOffset
			BindOnLocalHost: true,
			BindOnIP:        "",
		}
	}

	// Assign any open port when configured to use dynamic ports
	if o.DynamicPorts {
		if frontendPortOffset != 0 {
			svc.RPC.GRPCPort = o.portProvider.mustGetFreePort()
		}
		svc.RPC.MembershipPort = o.portProvider.mustGetFreePort()
	}

	// Optionally bind frontend to IPv4 address
	if frontendPortOffset == 0 && o.FrontendAddress != "" {
		svc.RPC.BindOnLocalHost = false
		svc.RPC.BindOnIP = o.FrontendAddress
	}

	return svc
}

```

File: internal/liteconfig/config.go [hunk 2]
```diff
@@ -53,7 +54,8 @@ func NewDefaultConfig() (*Config, error) {
 			Level:      "debug",
 			OutputFile: "",
 		})),
-		portProvider: &portProvider{},
+		portProvider:    &portProvider{},
+		FrontendAddress: broadcastAddress,
```
Comment (jlegrone) [inline internal/liteconfig/config.go:58]: Maybe this should just default to an empty string after we rename to `FrontendIP`, since this config option is mutually exclusive with binding on `localhost` according to https://github.com/temporalio/temporal/blob/6daf3b6fb631cb58ae9522d5413b727a03531973/common/config/config.go#L89-L94

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FROM golang:1.17
+
+WORKDIR /go/src/temporalite
+ADD . /go/src/temporalite/
+RUN go build -o /go/bin/temporalite /go/src/temporalite/cmd/temporalite
+
+VOLUME ["/temporalite"]
+
+EXPOSE 7233
+ENTRYPOINT ["/go/bin/temporalite", "start", "-f", "/temporalite/temporalite.db", "-n", "default", "-a" , "0.0.0.0"]
```
Comment (jlegrone) [inline Dockerfile:10]: Let's move this into a separate PR. I'd also love to use a multi-stage build to remove the go compiler dependencies from the resulting image.

Comment (jbreiding) [reply Dockerfile:10]: given my branch name I will keep this one for docker and submit another for frontend-ip-address.

File: cmd/temporalite/main.go [hunk 4]
```diff
@@ -28,11 +29,12 @@ var (
...
-	dbPathFlag    = "filename"
-	portFlag      = "port"
-	logFormatFlag = "log-format"
-	namespaceFlag = "namespace"
+	ephemeralFlag       = "ephemeral"
+	dbPathFlag          = "filename"
+	portFlag            = "port"
+	logFormatFlag       = "log-format"
+	namespaceFlag       = "namespace"
+	frontendAddressFlag = "frontend-address"
```
[the remark this answers is not in the document]
Comment (jbreiding) [reply cmd/temporalite/main.go]: Moved this and pulled in changes here
https://github.com/DataDog/temporalite/pull/17/files

File: internal/liteconfig/config.go (hunk 1 above)
Comment (jbreiding) [reply internal/liteconfig/config.go]: Moved this and pulled in changes here
https://github.com/DataDog/temporalite/pull/17/files

File: Dockerfile (hunk 3 above)
Comment (jbreiding) [reply Dockerfile]: This PR is now for the Dockerfile only.

Based on distress guidance I have converted to a multistage build.

**Comment by jbreiding** [comment]
This PR should be sequenced after #17

File: internal/liteconfig/config.go (hunk 2 above)
Comment (jbreiding) [reply internal/liteconfig/config.go]: Moved this and pulled in changes here
https://github.com/DataDog/temporalite/pull/17/files

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ADD . .
+RUN go mod download
+RUN go get -d -v ./...
+RUN go build -o /go/bin/temporalite ./cmd/temporalite
+
+FROM gcr.io/distroless/base-debian11
+
+COPY --from=builder /go/bin/temporalite /
+EXPOSE 7233
+ENTRYPOINT ["/temporalite", "start", "-f", "/tmp/temporalite.db"]
```
Comment (jlegrone) [inline Dockerfile:13]: Should we also bind to 0.0.0.0 for convenience?

**Review by jlegrone — APPROVED** [review]
@jbreiding I could also take a stab at this later, but would you be up for adding an e2e test to the Github CI workflow to ensure that the docker image can be successfully built and when executed exposes a healthy Temporal service endpoint?

File: Dockerfile (hunk 5 above)
Comment (jlegrone) [reply Dockerfile:13]: Or borrow from the upstream entrypoint script: https://github.com/temporalio/temporal/blob/master/docker/entrypoint.sh

Comment (jbreiding) [reply Dockerfile:13]: These are great ideas, I will see where I can find the time to sequence in the work.

Comment (jbreiding) [reply Dockerfile:13]: > Or borrow from the upstream entrypoint script: https://github.com/temporalio/temporal/blob/master/docker/entrypoint.sh

I am not sure that is possible using distro-less given there is no shell.

Alpine has its own issues given the differences between musl and glibc.

I am using a combo container for my testing is based on Debian bullseye and has a copy of tcl.
The ENTRYPOINT is easily overridden from docker cli or even docker-compose.

Comment (jbreiding) [reply Dockerfile]: This is updated now, a simple sane default.

Updated the PR comment to show execution commands.

**Comment by jbreiding** [comment]
> @jbreiding I could also take a stab at this later, but would you be up for adding an e2e test to the Github CI workflow to ensure that the docker image can be successfully built and when executed exposes a healthy Temporal service endpoint?

Perhaps in another PR.
This would likely entail _dockerizing_ the test worker this repo or using tctl and maybe combining with a docker-compose example.

**Review by jlegrone — APPROVED** [review]
